src/gluonts/nursery/MSDA/model/model.py

```diff
@@ -0,0 +1,1020 @@
+import os
+import torch
+import torch.nn as nn
+import torch.optim as optim
+import torch.nn.functional as F
+import torch.optim.lr_scheduler as lr_scheduler
+import numpy as np
+from model.modules import (
+    FeatureNet,
+    PredNet,
+    GNet,
+    ClassDiscNet,
+    CondClassDiscNet,
+    DiscNet,
+    GraphDNet,
+)
+import pickle
+from visdom import Visdom
+
+# ===========================================================================================================
+
+
+def to_np(x):
+    return x.detach().cpu().numpy()
+
+
+def to_tensor(x, device="cuda"):
+    if isinstance(x, np.ndarray):
+        x = torch.from_numpy(x).to(device)
+    else:
+        x = x.to(device)
+    return x
+
+
+def flat(x):
+    n, m = x.shape[:2]
+    return x.reshape(n * m, *x.shape[2:])
+
+
+def write_pickle(data, name):
```

Comment:
This function is repeated a couple of times. Can we make sure we have it written only once?